File: ReplaceConditionalLogicWithStrategyPeter/src/Loan.java
import java.util.ArrayList;
import java.util.Date;
import java.util.Iterator;

public class Loan {
    private static final int MILLIS_PER_DAY = 86400000;
    private static final int DAYS_PER_YEAR = 365;

    public final double commitment;
    public final double outstanding;
    public final int riskRating;
    public final Date maturity;
    public final Date expiry;
    public CapitalStrategy capitalStrategy;
    private ArrayList payments;
    private Date today;
    private Date start;

    private Loan(CapitalStrategy capitalStrategy, double commitment, double outstanding, int riskRating, Date maturity, Date expiry) {
        this.commitment = commitment;
        this.outstanding = outstanding;
        this.riskRating = riskRating;
        this.maturity = maturity;
        this.expiry = expiry;
        this.capitalStrategy = capitalStrategy;
        this.payments = new ArrayList();
        this.today = new Date();
        this.start = new Date();

        if (capitalStrategy == null) {
            if(expiry == null)
                this.capitalStrategy = new CapitalStrategyTermLoan();
            else if (maturity == null)
                this.capitalStrategy = new CapitalStrategyRevolver();
            else
                this.capitalStrategy = new CapitalStrategyRCTL();
        }
    }

    public double capital() {
        if (expiry == null && maturity != null)
            return commitment * duration() * riskFactor();
        if (expiry != null && maturity == null) {
            if (getUnusedPercentage() != 1.0)
                return commitment * getUnusedPercentage() * duration() * riskFactor();
            else
                return (outstandingRiskAmount() * duration() * riskFactor())
                        + (unusedRiskAmount() * duration() * unusedRiskFactor());
        }
        return 0.0;
    }

    private double getUnusedPercentage() {
        return 0;
    }

    private double outstandingRiskAmount() {
        return outstanding;
    }

    private double unusedRiskAmount() {
        return (commitment - outstanding);
    }

    public double duration() {
        if (expiry == null && maturity != null)
            return weightedAverageDuration();
        else if (expiry != null && maturity == null)
            return yearsTo(expiry);
        return 0.0;
    }

    private double weightedAverageDuration() {
        double duration = 0.0;
        double weightedAverage = 0.0;
        double sumOfPayments = 0.0;
        Iterator loanPayments = payments.iterator();
        while (loanPayments.hasNext()) {
            Payment payment = (Payment) loanPayments.next();
            sumOfPayments += payment.amount();
            weightedAverage += yearsTo(payment.date()) * payment.amount();
        }
        if (commitment != 0.0)
            duration = weightedAverage / sumOfPayments;
        return duration;
    }

    private double yearsTo(Date endDate) {
        Date beginDate = (today == null ? start : today);
        return ((endDate.getTime() - beginDate.getTime()) / MILLIS_PER_DAY) / DAYS_PER_YEAR;
    }

    private double riskFactor() {
        return RiskFactor.getFactors().forRating(riskRating);
    }

    private double unusedRiskFactor() {
        return UnusedRiskFactors.getFactors().forRating(riskRating);
    }
}

File: FormTemplateMethodRuby/lib/loan.rb
require_relative 'capital_strategy_term_loan'
require_relative 'capital_strategy_advised_line'
require_relative 'capital_strategy_revolver'

# Useful comment
class Loan

  attr_reader :expiry, :maturity, :commitment, :outstanding, :start, :risk_rating
  attr_accessor :unused_percentage

  def initialize(commitment, outstanding, start, expiry, maturity, risk_rating, capital_strategy)
    @start = start
    @expiry = expiry
    @maturity = maturity
    @commitment = commitment
    @outstanding = outstanding
    @risk_rating = risk_rating
    @capital_strategy = capital_strategy
  end

  def self.new_revolver(commitment, start, expiry, risk_rating)
    new(commitment, 0, start, expiry, nil, risk_rating, CapitalStrategyRevolver.new)
  end


  def self.new_term_loan(commitment, start, maturity, risk_rating)
    new(commitment, commitment, start, nil, maturity, risk_rating, CapitalStrategyTermLoan.new)
  end

  def self.new_advised_line(commitment, start, expiry, risk_rating)
    return nil if risk_rating > 3

    advised_line = new(commitment, 0, start, expiry, nil, risk_rating, CapitalStrategyAdvisedLine.new)
    advised_line.unused_percentage = 0.1
    advised_line
  end

  def capital
    @capital_strategy.capital(self)
  end

  def duration
    @capital_strategy.duration(self)
  end

  def outstanding_risk_amount
    7
  end

  def unused_risk_amount
    commitment - outstanding_risk_amount
  end

end